design/20210203.certificate-request-identity.md

```diff
@@ -0,0 +1,165 @@
+
+## Proposal
+
+Kubernetes does [not currently support immutable
+fields](https://github.com/kubernetes/enhancements/blob/8b9b994136371f1bc938aabf012f4c45535d684c/keps/sig-api-machinery/20190603-immutable-fields.md)
+for CRDs. The cert-manager webhook will be responsible for populating and
+enforcing identity fields which are present on the spec of `CertificateRequest`
+resources.
+
+The webhook will be responsible for enforcing the following during a CREATE
+operation:
+- No identity fields have been set by the user creating the resource
+- Set identity fields to exactly what is received from the API server in the
+    [UserInfo](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.20/#userinfo-v1-authentication-k8s-io)
+
+The webhook will also responsible for enforcing the following during an UPDATE
+operation:
+- No changes to identity fields are allowed to be made
+
+### API Changes
+
+In order to expose the identity of who created `CertificateRequest` resources,
+these resources must be updated to have parity with the upstream Kubernetes
+certificates API.
+This means that the `CertificateRequest` API type be updated to include the
+following fields in spec, for all API versions:
+
+```go
+	// Username contains the name of the user that created the CertificateRequest.
+	// Populated by the cert-manager webhook on creation and immutable.
+	// +optional
+	Username string `json:"username,omitempty"`
+	// UID contains the uid of the user that created the CertificateRequest.
+	// Populated by the cert-manager webhook on creation and immutable.
+	// +optional
+	UID string `json:"uid,omitempty"`
+	// Groups contains group membership of the user that created the CertificateRequest.
+	// Populated by the cert-manager webhook on creation and immutable.
+	// +listType=atomic
+	// +optional
+	Groups []string `json:"groups,omitempty"`
+	// Extra contains extra attributes of the user that created the CertificateRequest.
+	// Populated by the cert-manager webhook on creation and immutable.
+	// +optional
+	Extra map[string][]string `json:"extra,omitempty"`
+```
+
+### Upgrading
+
+Any `CertificateRequest`s that are present in the cluster at the time of
+upgrade, won't and will never have their identity fields populated. This should
+be acceptable.
+
+All subsequent `CertificateRequest`s created after this upgrade will have their
+identities populated.
+
+### Test Plan
+
+Unit tests will be created is ensure that the [properties](#proposal) are
+enforced.
+
+End to end tests will ensure that different users creating `CertificateRequest`
+resources have their respective identities populated in the fields.
+
+### Risks and Mitigations
+
+There are large security implications if the identity fields that other
+components rely on are wrong, or could be made fraudulent. Special care when
+testing needs to be given to ensure the properties described above are correct
+and enforced.
```

Comment:
May be worth calling out that the failure modes that _could_ occur would all be pre-tensed on the Mutating/validating webhook not being installed correctly (either the resources being removed, or the webhook being out of date etc).

Also, emphasis during the upgrade path on the fact that if the CRD is update to include these fields *before* the webhook is upgrade, there will be no restrictions on who can set those fields for that brief window as the apiserver will have no way to enforce those fields to not be set. I don't think we need to mitigate against this personally, but it's an edge-case to be aware of :)